icu-date.lua

```diff
@@ -50,9 +51,14 @@ local function string_to_uchar(str)
   return uchar
 end
 
+-- UChar is unsigned short. For different locales size of string
+-- in bytes could be greater than string length.
+-- For such cases we should reserve sizeof('UChar') * (length + 1)
+-- bytes.
+-- See https://github.com/unicode-org/icu/blob/2dce62892b6b00df02195a35ca09ebd8e082e80f/docs/userguide/format_parse/datetime/examples.md#c-5
```

Comment:
It should work in most cases, but I doubt that there are no opposite examples. Let me formalize the problem. We have given UTF-16 string. We should determine size of a byte buffer suitable to store a string, which is conversion of the original string to a system encoding. The system encoding is usually UTF-8, let's consider it as example.

The unicode symbols length in UTF-8 and UTF-16 based on [this](https://stackoverflow.com/a/5728481/1598057) SO answer:

| Range | UTF-16 | UTF-8 |
| ----- | ------ | ----- |
| U+0000-U+007F | 2 bytes | 1 byte |
| U+0080-U+07FF | 2 bytes | 2 bytes |
| U+0800-U+FFFF | 2 bytes | 3 bytes |
| U+10000-U+10FFFF | 4 bytes  | 4 bytes |

So, it seems, just twice amount of UTF-16 symbols not necessarily enough amount of bytes to represent the string in UTF-8. Don't know about other encodings.

I think we should use this 'same amount of bytes' buffer as default, but increase the buffer size if an overflow is detected. Is there an API to convert from ICU's internal representation to a system encoding, which checks buffer boundaries?

Reply:
I mean, the fix is meaningful (as it'll mostly work) and I'm ready to merge it, but if you know how to handle it for all cases (or curious and ready to investigate), it would be nice to resolve the problem entirely.

At least, please, update the commit/PR messages with appropriate info about the fix.

Reply:
> I think we should use this 'same amount of bytes' buffer as default, but increase the buffer size if an overflow is detected. Is there an API to convert from ICU's internal representation to a system encoding, which checks buffer boundaries?

I've failed to find such API. But there are only u_austrcpy and u_austrncpy functions for such purpose. https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ustring_8h.html

I agree with "we should determine size of a byte buffer suitable to store a string" and wrote about it [here](https://github.com/tarantool/icu-date/pull/23#discussion_r705463535). But I also failed to find such API. Internally inside u_austrcpy libicu doesn't use such function.

But here we have UChar that actually is unsigned short (2 bytes) and I guess in case we work with 4 byte chars we will use 2 UChars.
I've made a small experiment with 4 byte char:
```
str = '𐀀'
#'𐀀'
---
- 4
...
res = string_to_uchar(str)
ffi.C.u_strlen_62(res)         # In case of back conversion to default encoding we will allocate 4 byte buffer now
---
- 2
...
```

Reply:
There is no problem around the `U+10000-U+10FFFF` range.

```diff
diff --git a/icu-date.lua b/icu-date.lua
index 3c37c1a..2e364c5 100755
--- a/icu-date.lua
+++ b/icu-date.lua
@@ -50,6 +50,7 @@ local function string_to_uchar(str)
   call_fn("u_uastrcpy", uchar, str)
   return uchar
 end
+_M.string_to_uchar = string_to_uchar
 
 -- UChar is unsigned short. For different locales size of string
 -- in bytes could be greater than string length.
@@ -62,6 +63,7 @@ local function uchar_to_string(uchar)
   call_fn("u_austrcpy", str, uchar)
   return ffi.string(str)
 end
+_M.uchar_to_string = uchar_to_string
 
 local function check_self_type(self, func)
   assert(type(self) == "table", ("Use date:%s(...) instead of date.%s(...)"):format(func, func))
```

```yaml
tarantool> ffi.C.u_strlen(icu_date.string_to_uchar('\u{10000}')) * ffi.sizeof('UChar')
---
- 4
...

tarantool> #icu_date.uchar_to_string(icu_date.string_to_uchar('\u{10000}'))
---
- 4
...
```

Let's look at `U+0800-U+FFFF`:

```yaml
tarantool> ffi.C.u_strlen(icu_date.string_to_uchar('\u{0800}')) * ffi.sizeof('UChar')
---
- 2
...

tarantool> #icu_date.uchar_to_string(icu_date.string_to_uchar('\u{0800}'))
---
- 3
...
```

Two bytes in UTF-16 (one UChar). Three bytes in UTF-8.

Reply:
Ok. Thanks a lot for explanation. I've update my patch and add a test with u0800. Currently is should be better.